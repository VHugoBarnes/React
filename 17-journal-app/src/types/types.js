export const types = {
    // Types para auth
    login: '[Auth] Login',
    logout: '[Auth] Logout',

    // Types para ui
    uiSetError: '[UI] Set Error',
    uiRemoveError: '[UI] Remove Error',
    uiStartLoading: '[UI] Start Loading',
    uiFinishLoading: '[UI] Finish Loading',

};
