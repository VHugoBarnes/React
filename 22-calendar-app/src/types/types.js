export const types = {
    // UI
    uiOpenModal: '[UI] Open modal',
    uiCloseModal: '[UI] Close modal',

    // Calendar
    eventAddNew: '[Event] Add new',
    eventSetActive: '[Event] Set active',
    eventClearActiveEvent: '[Event] Clear active event',
    eventUpdated: '[Event] Event updated',
    eventDeleted: '[Event] Event deleted'
}