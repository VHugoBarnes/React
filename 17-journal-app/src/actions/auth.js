import { types } from "../types/types";
import { googleAuthProvider, firebase } from '../firebase/firebaseConfig';
import { startLoading, finishLoading } from "./ui";

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch( startLoading() );
        // auth()
        firebase.auth().signInWithEmailAndPassword(email, password)
            .then( ({user}) => {
                dispatch(
                    login(user.uid, user.displayName)
                );
                dispatch( finishLoading() );
            }).catch( e => {
                console.log(e);
                dispatch( finishLoading() );
            })
    }
};

export const startRegisterWithEmailPasswordName = ( username, email, password ) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then( async({user}) => {
                await user.updateProfile({ displayName: username });
                console.log(user);
                // dispatch(
                //     login(user.uid, user.displayName)
                // );
            }).catch( e => {
                console.log(e);
            });
    }
};

export const startGoogleLogin = () => {
    return ( dispatch ) => {
        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( ({user}) => {
                dispatch(
                    login(user.uid, user.displayName)
                );
            });
    }
};

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});

export const startLogout = () => {
    return async(dispatch) => {
        // Hacemos logout desde firebase
        await firebase.auth().signOut();
        // Borramos el uid de la memoria de la pÃ¡gina
        dispatch(logout());
    }
};

export const logout = () => ({
        type: types.logout
})

