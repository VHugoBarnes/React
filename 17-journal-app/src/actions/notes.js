import Swal from 'sweetalert2';
import { db } from "../firebase/firebaseConfig";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";
import { fileUpload } from '../helpers/fileUpload';

// Preset name for cloudinary: react-journal

export const startNewNote = () => {

    // Con el getState podemos acceder al estado actual de los reducers
    return async(dispatch, getState) => {
        // Sacamos del store el reducer auth y su valor uid.
        const { uid } = getState().auth;
        // Creamos el objeto
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime(),
        }
        // Llamamos a un path en firestore para guardar la información en una colección
        const doc = await db.collection(`${uid}/journal/notes`).add( newNote );
        dispatch( activeNote(doc.id, newNote) );
    }
}

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = (uid) => {
    return async( dispatch ) => {
        const notes = await loadNotes( uid );
        dispatch(setNotes(notes));
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes,
})

export const startSaveNote = ( note ) => {
    return async(dispatch, getState) => {
        // Sacamos del store el reducer auth y su valor uid.
        const { uid } = getState().auth;

        if(!note.url) {
            delete note.url;
        }

        // Extraemos la nota (una copia) a una variable
        const noteToFirestore = {...note};
        // Se elimina el id porque no se puee actualizar
        delete noteToFirestore.id;

        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);

        dispatch(refreshNote(note.id, noteToFirestore));
        Swal.fire('Saved', note.title, 'success');
    }
}

export const refreshNote = (id, note) => ({
    type: types.notesUpdate,
    payload: {
        id, 
        note: {id, ...note}
    }
})

export const startUploading = (file) => {
    return async(dispatch, getState) => {
        const {active: activeNote} = getState().notes;
        // console.log(file);
        // console.log(activeNote);
        const fileUrl = await fileUpload(file);
        console.log(fileUrl);
    }
}
