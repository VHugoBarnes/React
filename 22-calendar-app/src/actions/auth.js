import Swal from "sweetalert2";
import { fetchSinToken } from "../helpers/fetch";
import { types } from "../types/types";

/**
 * Acción asíncrona que hace una petición al backend
 * para iniciar sesión con los datos mandados por parámetro.
 * 
 * Ocupa thunk, por eso retorna una función (asíncrona).
 * 
 * @param {string} email el correo pasado del formulario
 * @param {string} password la contraseña pasada por el formulario
 */
export const startLogin = ( email, password ) => {
    return async( dispatch ) => {
        
        // Llamada a la API
        const resp = await fetchSinToken( 'auth', { email, password }, 'POST' );

        // La respuesta de la API
        const body = await resp.json();
        
        // La API manda un objeto con la propiedad 'ok'
        if ( body.ok ) {
            // Se almacena en el localstorage el JWT generado.
            // Como no es información sensible lo podemos guardar ahí sin problems ;)
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            // Guardar en el store
            dispatch( login({
                uid: body.uid,
                body: body.name
            }) )
        } else { // En caso de errores :o
            Swal.fire('Error', body.msg, 'error');
        }

    }
}

export const startRegister = ( email, password, name ) => {
    return async( dispatch ) => {
        // Llamada a la API
        const resp = await fetchSinToken( 'auth/new', { email, password, name }, 'POST' );

        // La respuesta de la API
        const body = await resp.json();
        
        // La API manda un objeto con la propiedad 'ok'
        if ( body.ok ) {
            // Se almacena en el localstorage el JWT generado.
            // Como no es información sensible lo podemos guardar ahí sin problems ;)
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            // Guardar en el store
            dispatch( login({
                uid: body.uid,
                body: body.name
            }) )
        } else { // En caso de errores :o
            Swal.fire('Error', body.msg, 'error');
        }
    }
}

/**
 * Acción síncrona para hacer un login en la web.
 * Básicamente lo que hace es guardar en el store el usuario pasado por param.
 * 
 * @param {object} user El objeto que contiene la información del usuario para almacenarlo en el store
 */
const login = ( user ) => ({
    type: types.authLogin,
    payload: user
});