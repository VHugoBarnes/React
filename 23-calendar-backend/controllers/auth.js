const express = require('express');
const bcrypt = require('bcryptjs');
const Usuario = require('../models/Usuario');
const { generarJWT } = require('../helpers/jwt');

const createUser = async(req, res = express.response ) => {

    const { email, password } = req.body;

    try {
        let usuario = await Usuario.findOne({email});
        
        if (usuario) {
            return res.status(400).json({
                ok: false,
                msg: 'El usuario ya existe con ese correo',
            });
        }

        usuario = new Usuario(req.body);

        // Encriptar contraseña
        const salt = bcrypt.genSaltSync();
        usuario.password = bcrypt.hashSync( password, salt );

        await usuario.save();

        // Generar JWT (JSON Web Token)
        const token = await generarJWT( usuario.id, usuario.name );

        res.status(201).json({
            ok: true,
            uid: usuario.id,
            name: usuario.name,
            token
        });
    } catch (error) {
        console.error(error);
        res.status(500).json({
            ok: false,
            msg: 'Por favor hable con el administrador'
        });
    }
};

const userLogin = async(req, res = express.response ) => {

    const { email, password } = req.body;

    try {
        const usuario = await Usuario.findOne({email});

        if (!usuario) {
            return res.status(400).json({
                ok: false,
                msg: 'El usuario no existe con ese email',
            });
        }

        // Confirmar las contraseñas
        const validPassword = bcrypt.compareSync(password, usuario.password);

        if ( !validPassword ) {
            return res.status(400).json({
                ok: false,
                msg: 'Password incorrecto'
            });
        }

        // Generar el JWT (JSON Web Token)
        const token = await generarJWT( usuario.id, usuario.name );

        res.status(200).json({
            ok: true,
            uid: usuario.id,
            name: usuario.name,
            token
        });

    } catch (error) {
        console.error(error);
        res.status(500).json({
            ok: false,
            msg: 'Por favor hable con el administrador'
        });
    }
};

const tokenRenew = async(req, res = express.response ) => {

    const uid = req.uid;
    const name = req.name;

    // Generar un nuevo JWT y retornarlo en esta petición
    const token = await generarJWT( uid, name );;

    res.json({
        ok: true,
        msg: 'renew',
        token
    });
};

module.exports = {
    createUser,
    userLogin,
    tokenRenew,
}