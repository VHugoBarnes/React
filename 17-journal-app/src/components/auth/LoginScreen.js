import React from 'react';
import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { login, startLoginEmailPassword } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';

export const LoginScreen = () => {

    /**
     * Con este hook podemos hacer un dispatch donde nos encontremos
     * obvio siempre y cuando nos encontremos dentro de la app.
     * 
     * Sacado de la documentación oficial (https://react-redux.js.org/api/hooks#usedispatch):
     * Este hook retorna una referencia de la función `dispatch` del store de Redux
     * 
     * Esto quiere decir que se toma referencia de la función dispatch del store que a su vez
     * está 'linkeado' con el reducer que hicimos (authReducer), que no es más que un 
     * reducer común y corriente.
     */
    const dispatch = useDispatch();

    const [values, handleInputChange, reset] = useForm({
        email: 'victor@gmail.com',
        password: '123456',
    });

    const { email, password } = values;
    
    const handleLogin = (e) => {
        e.preventDefault();
        /**
         * Invocamos el dispatch, en este caso, login
         * este método devuelve {
         *      type: types.login,
         *      payload: {uid, displayName}
         * }
         */
        dispatch( startLoginEmailPassword(email, password) );
    }

    return (
        <>
            <h3 className="auth__title">Login</h3>
            
            <form onSubmit={ handleLogin }>
                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={ email }
                    onChange={ handleInputChange }
                />
                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={ password }
                    onChange={ handleInputChange }
                />
                <button
                    type="submit"
                    className="btn btn-primary btn-block"
                >
                    Login
                </button>
                <div className="auth__social-networks">
                    <p>Login with Social Networks</p>
                    <div 
                        className="google-btn"
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>
                <Link 
                    to="/auth/register"
                    className="link"
                >
                    Create new account
                </Link>
            </form>
        </>
    );
};
