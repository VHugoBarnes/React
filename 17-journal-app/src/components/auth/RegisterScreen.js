import React from 'react';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';

import { useForm } from '../../hooks/useForm';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

export const RegisterScreen = () => {

    const dispatch = useDispatch();
    const { msgError } = useSelector( state => state.ui);

    // Inicializando nuestro hook para manejar formularios
    const [values, handleInputChange] = useForm({
        username: 'VÃ­ctor',
        email: 'hugo@mail.com',
        password: '123456',
        confirmPassword: '123456',
    });

    // Desestructuramos el objeto values que regresa el `useForm`
    const { username, email, password, confirmPassword } = values;
    // Procedemos a mandarlo como value dentro de los input.

    // Handler para manejar el registro
    // DE MOMENTO: imprimir en consola los values de los input
    const handleRegister = ( e ) => {
        // Prevenimos el comportamiento por defecto al hacer submit
        e.preventDefault();
        
        if( isFormValid() ){
            dispatch(startRegisterWithEmailPasswordName(username, email, password));
        }
    }

    // Comprobar si el formulario es valido.
    const isFormValid = () => {
        if(username.trim().length === 0) {
            console.log('Name is required');
            dispatch(setError('Name is required'));
            return false;
        } else if( !validator.isEmail(email) ) {
            console.log('Email is not correct');
            dispatch(setError('Email is not correct'));
            return false;
        } else if( password !== confirmPassword || password.length < 5){
            console.log('Password should be at least 6 characters and match');
            dispatch(setError('Password should be at least 6 characters and match'));
            return false;
        }
        dispatch(removeError());
        return true;
    }

    return (
        <>
            <h3 className="auth__title">Register</h3>
            
            <form onSubmit={ handleRegister }>

                {
                    msgError !== null 
                        && <div className="auth__alert-error">
                                { msgError }
                            </div>
                }

                <input
                    type="text"
                    placeholder="Username"
                    name="username"
                    className="auth__input"
                    autoComplete="off"
                    value={ username }
                    onChange={ handleInputChange }
                />
                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={ email }
                    onChange={ handleInputChange }
                />
                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={ password }
                    onChange={ handleInputChange }
                />
                <input
                    type="password"
                    placeholder="Confirm Password"
                    name="confirmPassword"
                    className="auth__input"
                    value={ confirmPassword }
                    onChange={ handleInputChange }
                />
                <button
                    type="submit"
                    className="btn btn-primary btn-block mb-5"
                >
                    Register
                </button>
                <Link
                    to="/auth/login"
                    className="link"
                >
                    Already registered?
                </Link>
            </form>
        </>
    );
};
