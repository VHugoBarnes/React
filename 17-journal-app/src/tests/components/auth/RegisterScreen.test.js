import '@testing-library/jest-dom';
import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
// Librerías para hacer test con thunk
import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';
import { MemoryRouter } from 'react-router-dom';

import { startRegisterWithEmailPasswordName } from '../../../actions/auth';
import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/types';

// Configuración de thunk antes de iniciar los tests
const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
};
// El estado del store en este instante
let store = mockStore(initState);
// Como todas mis acciones son síncronas no hay necesidad de hacer un mock
// store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <RegisterScreen/>
        </MemoryRouter>
    </Provider>
);

describe('Pruebas en <RegisterScreen/>', () => {
    
    test('Debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot()
    });

    test('Debe de hacer el dispatch de la acción respectiva', () => {
        const emailField = wrapper.find('input[name="email"]');

        // Hacemos un cambio, simulamos un cambio
        // como estamos trabajando con nuestro customHook
        // necesitamos poner el target para hacer referencia a qué campo
        // se va a cambiar.
        emailField.simulate('change', {
            target: {
                value: '',
                name: 'email'
            }
        });
        
        // Simulamos dar enter al formulario.
        // al no tener nada en el email, debe de disparar la acción correspondiente
        wrapper.find('form').simulate('submit',{
            preventDefault(){}
        });

        const actions = store.getActions();
        expect(actions[0]).toEqual({
            type: types.uiSetError,
            payload: 'Email is not correct'
        });
    });

    test('Debe de mostrar la caja de alerta con el error', () => {
        const initState = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'email no es correcto'
            }
        };
        // El estado del store en este instante
        const store = mockStore(initState);

        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen/>
                </MemoryRouter>
            </Provider>
        );

        expect(wrapper.find('.auth__alert-error').exists()).toBeTruthy();
        expect(wrapper.find('.auth__alert-error').text()).toBe(initState.ui.msgError);
    });
    
});
