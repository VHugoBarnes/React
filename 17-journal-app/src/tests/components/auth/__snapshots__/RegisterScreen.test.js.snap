// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <RegisterScreen/> Debe de mostrarse correctamente 1`] = `
<div
  className="animate__animated animate__fadeIn animate__faster"
>
  <h3
    className="auth__title"
  >
    Register
  </h3>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="username"
      onChange={[Function]}
      placeholder="Username"
      type="text"
      value="VÃ­ctor"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="hugo@mail.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      className="auth__input"
      name="confirmPassword"
      onChange={[Function]}
      placeholder="Confirm Password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>
</div>
`;
